Allow enforcing single/double-quotes without preventing use of template literals

How it worked in ESLint 1.0.0 worked perfectly. I could enforce single-quotes everywhere `'quotes': [2, 'single']` and still use template literals when convenient, e.g. if the content had single-quotes so not having to escape it. With ESLint 2.0.0 template literals are now awkwardly forced into the rule and there's no way to enforce single-quotes without also preventing use of template literals when they're not strictly required. I don't want to use double-quotes to prevent having to escape strings containing single-quotes. I want to use template literals, or rather, I already do, and this change makes everything break for me...

I've read both https://github.com/eslint/eslint/issues/5147 and https://github.com/eslint/eslint/issues/3090, but I disagree with the conclusion and I think they're missing this use-case.

To be clear, this used to pass:

```
~/dev
❯ eslint --version                                                                           
v2.0.0
                                                                                                                                               
~/dev
❯ echo "var foo = \`bar 'foo' baz\`;" | eslint --stdin --env=es6 --rule='quotes: [2, single]'

<text>
  1:11  error  Strings must use singlequote  quotes

✖ 1 problem (1 error, 0 warnings)
```

I would like an option to allow this functionality.

--- a/docs/rules/quotes.md
+++ b/docs/rules/quotes.md
@@ -23,16 +23,19 @@ This rule is aimed at ensuring consistency of string quotes and as such will rep
 The rule configuration takes up to two options:
 
 1. The first option is `"double"`, `"single"` or `"backtick"` for double-quotes, single-quotes or backticks respectively. The default is `"double"`.
-1. The second option is the `"avoid-escape"` flag. When using `"avoid-escape"`, this rule will not report a problem when a string is using single-quotes or double-quotes so long as the string contains a quote that would have to be escaped otherwise. For example, if you specify `"double"` and `"avoid-escape"`, the string `'He said, "hi!"'` is not considered a problem because using double quotes for that string would require escaping the double quotes inside of the string. This option is off by default.
+1. The second option takes two options:
+    1. `"avoidEscape"`: When using `"avoidEscape"`, this rule will not report a problem when a string is using single-quotes or double-quotes so long as the string contains a quote that would have to be escaped otherwise. For example, if you specify `"double"` and `"avoidEscape"`, the string `'He said, "hi!"'` is not considered a problem because using double quotes for that string would require escaping the double quotes inside of the string. This option is off by default.
+    1. `"allowTemplateLiterals"`: when using `"allowTemplateLiterals"`, this rule will not report a problem when a string is using backticks and option one is either `"double"` or `"single"`.
 
-When using `"single"` or `"double"`, template literals that don't contain a substitution, don't contain a line break and aren't tagged templates, are flagged as problems, even with the `"avoid-escape"` option.
-
+When using `"single"` or `"double"`, template literals that don't contain a substitution, don't contain a line break and aren't tagged templates, are flagged as problems, even with the `"avoidEscape"` option. However they are not problems when `"allowTemplateLiterals"` is used.
 
 Configuration looks like this:
 
 ```js
-[2, "single", "avoid-escape"]
+[2, "single", {"avoidEscape": true, "allowTemplateLiterals": true}]
 ```
+
+**Deprecation notice**: The `avoid-escape` option is a deprecated syntax and you should use the object form instead.
 
 The following patterns are considered problems:
 
@@ -51,14 +54,14 @@ var unescaped = "a string containing 'single' quotes";
 ```
 
 ```js
-/*eslint quotes: ["error", "double", "avoid-escape"]*/
+/*eslint quotes: ["error", "double", {"avoidEscape": true}]*/
 
 var single = 'single';
 var single = `single`;
 ```
 
 ```js
-/*eslint quotes: ["error", "single", "avoid-escape"]*/
+/*eslint quotes: ["error", "single", {"avoidEscape": true}]*/
 
 var double = "double";
 var double = `double`;
@@ -73,7 +76,7 @@ var unescaped = 'a string containing `backticks`';
 ```
 
 ```js
-/*eslint quotes: ["error", "backtick", "avoid-escape"]*/
+/*eslint quotes: ["error", "backtick", {"avoidEscape": true}]*/
 
 var single = 'single';
 var double = "double";
@@ -99,15 +102,29 @@ var backtick = `back${x}tick`; // backticks are allowed due to substitution
 ```
 
 ```js
-/*eslint quotes: ["error", "double", "avoid-escape"]*/
+/*eslint quotes: ["error", "double", {"avoidEscape": true}]*/
 
 var single = 'a string containing "double" quotes';
 ```
 
 ```js
-/*eslint quotes: ["error", "single", "avoid-escape"]*/
+/*eslint quotes: ["error", "single", {"avoidEscape": true}]*/
 
 var double = "a string containing 'single' quotes";
+```
+
+```js
+/*eslint quotes: ["error", "double", {"allowTemplateLiterals": true}]*/
+
+var single = 'single';
+var single = `single`;
+```
+
+```js
+/*eslint quotes: ["error", "single", {"allowTemplateLiterals": true}]*/
+
+var double = "double";
+var double = `double`;
 ```
 
 ```js
@@ -118,7 +135,7 @@ var backtick = `backtick`;
 ```
 
 ```js
-/*eslint quotes: ["error", "backtick", "avoid-escape"]*/
+/*eslint quotes: ["error", "backtick", {"avoidEscape": true}]*/
 
 var double = "a string containing `backtick` quotes"
 ```

--- a/lib/rules/quotes.js
+++ b/lib/rules/quotes.js
@@ -78,8 +78,15 @@ module.exports = function(context) {
 
     var quoteOption = context.options[0],
         settings = QUOTE_SETTINGS[quoteOption || "double"],
-        avoidEscape = context.options[1] === AVOID_ESCAPE,
+        options = context.options[1],
+        avoidEscape = options && options.avoidEscape === true,
+        allowTemplateLiterals = options && options.allowTemplateLiterals === true,
         sourceCode = context.getSourceCode();
+
+    // deprecated
+    if (options === AVOID_ESCAPE) {
+        avoidEscape = true;
+    }
 
     /**
      * Determines if a given node is part of JSX syntax.
@@ -197,7 +204,7 @@ module.exports = function(context) {
         "TemplateLiteral": function(node) {
 
             // If backticks are expected or it's a tagged template, then this shouldn't throw an errors
-            if (quoteOption === "backtick" || node.parent.type === "TaggedTemplateExpression") {
+            if (allowTemplateLiterals || quoteOption === "backtick" || node.parent.type === "TaggedTemplateExpression") {
                 return;
             }
 
@@ -222,6 +229,22 @@ module.exports.schema = [
         "enum": ["single", "double", "backtick"]
     },
     {
-        "enum": ["avoid-escape"]
+        "anyOf": [
+            {
+                "enum": ["avoid-escape"]
+            },
+            {
+                "type": "object",
+                "properties": {
+                    "avoidEscape": {
+                        "type": "boolean"
+                    },
+                    "allowTemplateLiterals": {
+                        "type": "boolean"
+                    }
+                },
+                "additionalProperties": false
+            }
+        ]
     }
 ];

--- a/tests/lib/rules/quotes.js
+++ b/tests/lib/rules/quotes.js
@@ -24,17 +24,17 @@ ruleTester.run("quotes", rule, {
         { code: "var foo = \"bar\";", options: ["double"] },
         { code: "var foo = 1;", options: ["single"] },
         { code: "var foo = 1;", options: ["double"] },
-        { code: "var foo = \"'\";", options: ["single", "avoid-escape"] },
-        { code: "var foo = '\"';", options: ["double", "avoid-escape"] },
+        { code: "var foo = \"'\";", options: ["single", { avoidEscape: true }] },
+        { code: "var foo = '\"';", options: ["double", { avoidEscape: true }] },
         { code: "var foo = <div>Hello world</div>;", options: ["single"], parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } } },
         { code: "var foo = <div id=\"foo\"></div>;", options: ["single"], parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } } },
         { code: "var foo = <div>Hello world</div>;", options: ["double"], parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } } },
-        { code: "var foo = <div>Hello world</div>;", options: ["double", "avoid-escape"], parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } } },
+        { code: "var foo = <div>Hello world</div>;", options: ["double", { avoidEscape: true }], parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } } },
         { code: "var foo = `bar`;", options: ["backtick"], parserOptions: { ecmaVersion: 6 }},
         { code: "var foo = `bar 'baz'`;", options: ["backtick"], parserOptions: { ecmaVersion: 6 }},
         { code: "var foo = `bar \"baz\"`;", options: ["backtick"], parserOptions: { ecmaVersion: 6 }},
         { code: "var foo = 1;", options: ["backtick"]},
-        { code: "var foo = \"a string containing `backtick` quotes\";", options: ["backtick", "avoid-escape"] },
+        { code: "var foo = \"a string containing `backtick` quotes\";", options: ["backtick", { avoidEscape: true }] },
         { code: "var foo = <div id=\"foo\"></div>;", options: ["backtick"], parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } } },
         { code: "var foo = <div>Hello world</div>;", options: ["backtick"], parserOptions: { ecmaVersion: 6, ecmaFeatures: { jsx: true } }},
 
@@ -42,6 +42,11 @@ ruleTester.run("quotes", rule, {
         { code: "var foo = `back\ntick`;", options: ["single"], parserOptions: { ecmaVersion: 6 }},
         { code: "var foo = `back${x}tick`;", options: ["double"], parserOptions: { ecmaVersion: 6 }},
         { code: "var foo = tag`backtick`;", options: ["double"], parserOptions: { ecmaVersion: 6 }},
+
+        // Backticks are also okay if allowTemplateLiterals
+        { code: "var foo = `bar 'foo' baz` + 'bar';", options: ["single", { allowTemplateLiterals: true }], parserOptions: { ecmaVersion: 6 }},
+        { code: "var foo = `bar 'foo' baz` + \"bar\";", options: ["double", { allowTemplateLiterals: true }], parserOptions: { ecmaVersion: 6 }},
+        { code: "var foo = `bar 'foo' baz` + `bar`;", options: ["backtick", { allowTemplateLiterals: true }], parserOptions: { ecmaVersion: 6 }},
 
         // `backtick` should not warn the directive prologues.
         { code: "\"use strict\"; var foo = `backtick`;", options: ["backtick"], parserOptions: { ecmaVersion: 6 }},
@@ -111,19 +116,31 @@ ruleTester.run("quotes", rule, {
         {
             code: "var foo = \"bar\";",
             output: "var foo = 'bar';",
-            options: ["single", "avoid-escape"],
+            options: ["single", { avoidEscape: true }],
             errors: [{ message: "Strings must use singlequote.", type: "Literal" }]
         },
         {
             code: "var foo = 'bar';",
             output: "var foo = \"bar\";",
-            options: ["double", "avoid-escape"],
+            options: ["double", { avoidEscape: true }],
             errors: [{ message: "Strings must use doublequote.", type: "Literal" }]
         },
         {
             code: "var foo = '\\\\';",
             output: "var foo = \"\\\\\";",
-            options: ["double", "avoid-escape"],
+            options: ["double", { avoidEscape: true }],
+            errors: [{ message: "Strings must use doublequote.", type: "Literal" }]
+        },
+        {
+            code: "var foo = \"bar\";",
+            output: "var foo = 'bar';",
+            options: ["single", { allowTemplateLiterals: true }],
+            errors: [{ message: "Strings must use singlequote.", type: "Literal" }]
+        },
+        {
+            code: "var foo = 'bar';",
+            output: "var foo = \"bar\";",
+            options: ["double", { allowTemplateLiterals: true }],
             errors: [{ message: "Strings must use doublequote.", type: "Literal" }]
         },
         {
@@ -147,13 +164,13 @@ ruleTester.run("quotes", rule, {
         {
             code: "var foo = \"bar\";",
             output: "var foo = `bar`;",
-            options: ["backtick", "avoid-escape"],
+            options: ["backtick", { avoidEscape: true }],
             errors: [{ message: "Strings must use backtick.", type: "Literal" }]
         },
         {
             code: "var foo = 'bar';",
             output: "var foo = `bar`;",
-            options: ["backtick", "avoid-escape"],
+            options: ["backtick", { avoidEscape: true }],
             errors: [{ message: "Strings must use backtick.", type: "Literal" }]
         },
 

